src/balancer/src/lib.rs

```diff
@@ -0,0 +1,466 @@
+    #[tracing::instrument(level = "debug", skip_all)]
+    async fn run<'a, A>(
+        conn: &'a mut FramedConn<A>,
+        version: i32,
+        params: BTreeMap<String, String>,
+        resolver: &Resolver,
+        tls_mode: Option<TlsMode>,
+    ) -> Result<(), io::Error>
+    where
+        A: AsyncRead + AsyncWrite + AsyncReady + Send + Sync + Unpin,
+    {
+        if version != VERSION_3 {
+            return conn
+                .send(ErrorResponse::fatal(
+                    SqlState::SQLSERVER_REJECTED_ESTABLISHMENT_OF_SQLCONNECTION,
+                    "server does not support the client's requested protocol version",
+                ))
+                .await;
+        }
+
+        let Some(user) = params.get("user") else {
+            return conn
+                .send(ErrorResponse::fatal(
+                    SqlState::SQLSERVER_REJECTED_ESTABLISHMENT_OF_SQLCONNECTION,
+                    "user parameter required",
+                ))
+                .await;
+        };
+
+        if let Err(err) = conn.inner().ensure_tls_compatibility(&tls_mode) {
+            return conn.send(err).await;
+        }
+
+        let resolved = match resolver.resolve(conn, user).await {
+            Ok(v) => v,
+            Err(err) => {
+                return conn
+                    .send(ErrorResponse::fatal(
+                        SqlState::INVALID_PASSWORD,
+                        err.to_string(),
+                    ))
+                    .await;
+            }
+        };
+
+        if let Err(err) = Self::stream(conn, resolved.addr, resolved.password, params).await {
+            return conn
+                .send(ErrorResponse::fatal(
+                    SqlState::INVALID_PASSWORD,
+                    err.to_string(),
+                ))
+                .await;
+        }
+
+        Ok(())
+    }
+
+    #[tracing::instrument(level = "debug", skip_all)]
+    async fn stream<'a, A>(
+        conn: &'a mut FramedConn<A>,
+        envd_addr: SocketAddr,
+        password: Option<String>,
+        params: BTreeMap<String, String>,
+    ) -> Result<(), anyhow::Error>
+    where
+        A: AsyncRead + AsyncWrite + AsyncReady + Send + Sync + Unpin,
+    {
+        let client_stream = conn.inner_mut();
+        let mut mz_stream = TcpStream::connect(envd_addr).await?;
+        let mut buf = BytesMut::new();
+
+        // Send initial startup and password messages.
+        let startup = FrontendStartupMessage::Startup {
+            version: VERSION_3,
+            params,
+        };
+        startup.encode(&mut buf)?;
+        if let Some(password) = password {
+            let password = FrontendMessage::Password { password };
+            password.encode(&mut buf)?;
+        }
+        mz_stream.write_all(&buf).await?;
+
+        // Now blindly shuffle bytes back and forth until closed.
+        // TODO: Limit total memory use.
+        tokio::io::copy_bidirectional(client_stream, &mut mz_stream).await?;
```

Comment:
This is super neat! I didn't know this function existed. In the future I feel like we might want some logging and/or retry logic here though? Unless the retry logic is handled at a different layer already

Reply:
Yes, this is all coming in a future PR.